splicemachine/spliceengine
# DB-11905/12121 Create a channel between StreamListener and ResultStreamer to pause a streaming if overloaded

## Short Description
The additional channel between StgreamListener and StreamableRDD to manage throttling 

## Long Description
The throttling mechanism we have currently, see StreamableRDD.submit(), is to run 2 concurrent jobs at a time on the spark side, each sending rows from 2 partitions to the region server. If the buffer for a particular partition is full, it would cause back-pressure, and prevent Spark from sending more rows for that partition. However, if the number of rows in the partition is smaller than the buffer size, then there is another mechanism to hold Spark from sending all partitions to the region server. 
There is additional channel from StreamableRDD to StreamListener. It uses the same hostname/port, which is passed with submit command and used to connect from ResultStreamer.
When the StreamListener has more than 3 * DEFAULT_PARALLEL_PARTITIONS in the queue, it sends the message PauseStream to StreamableRDD to pause to submit new partitions to Spark. If the consumption starts again, StreamListener sends the message ContinueStream to resume the partions submitting.

The functionality can be disabled by configuration parameter **spark.result.streaming.throttle.enabled**, by default is true.

## How to test
```
create table a ( i int);
insert into a values 1,2,3;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
insert into a select i + (select count(*) from a) from a;
```

```
get cursor c as 'select * from a --splice-properties useSpark=true
order by i';
next c;

ps -ef | grep 'com.splicemachine.test.SpliceTestPlatform' | less

jmap -histo:live [PID] | grep 'PartitionState'

```
It should not be bigger than 12 on the standalone cluster
